### PR TITLE
feat: allow access to host CSSStyleDeclaration via `this.style`

### PR DESCRIPTION
**Pull request by le-incroyable1-dev**
## Details
Fixing #3999 

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.
* Host CSSStyleDeclaration can be accessed for Light DOM Components directly through .style

## Does this pull request introduce an observable change?

* ⚠️ Yes, it does include an observable change. 
*  this.style is no longer undefined

## GUS work item
Not Applicable

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/base-lightning-element.ts [hunk 1]
```diff
@@ -774,6 +776,11 @@ LightningElement.prototype = {
         return renderer.getTagName(elm);
     },
 
+    get style() {
+        const { elm } = getAssociatedVM(this);
+        return elm.style;
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/base-lightning-element.ts:781]: All DOM invocations must go through the `renderer` API. Please see `tagName` above for an example of how to do this. I.e., you must use the `getStyle` method you defined in `renderer/index.ts` in `@lwc/engine-dom`.

(This is a requirement due to how LWC integrates with Lightning Web Security on the Salesforce platform. The `style` call must be sandboxed by LWS.)

File: packages/@lwc/integration-karma/test/component/LightningElement.style/index.spec.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+it('should return the CSSStyleDeclaration of Host Element', () => {
+    const elm = createElement('x-test', { is: Test });
+    expect(elm.test).toEqual('X-TEST');
+});
+
+it('should be overrideable', () => {
+    const elm = createElement('x-override', { is: Override });
+    expect(elm.tagName).toEqual('x-foo');
+});
```
Comment (nolanlawson) [inline packages/@lwc/integration-karma/test/component/LightningElement.style/index.spec.js:14]: These tests make sense for `tagName`, but they don't verify the `style` at all. A better test would be to confirm that, for example `elm.style.color = 'red'` sets the `computedStyle` to be the color red.

File: playground/src/modules/x/counter/counter.html [hunk 3]
```diff
@@ -1,5 +1,5 @@
 <template>
-    Counter: {counter}
-    <button onclick={decrement}>-</button>
-    <button onclick={increment}>+</button>
-</template>
+  Counter: {counter}
+  <button onclick={decrement}>-</button>
+  <button onclick={increment}>+</button>
+</template>
```
Comment (nolanlawson) [inline playground/src/modules/x/counter/counter.html:5]: This change seems unrelated to this PR?

File: packages/@lwc/engine-server/src/__tests__/fixtures/style/expected.html [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<x-style>
+    <template shadowrootmode="open">
+    <span>
+      X-STYLE
```
Comment (nolanlawson) [inline packages/@lwc/engine-server/src/__tests__/fixtures/style/expected.html:4]: This doesn't seem to be the correct expected value. This is the `tagName`, not the style.

For `engine-server`, I think it would be fine to just throw an "unsupported method" error for the case of `getStyle`. It doesn't make much sense to me to be able to manipulate the `style` object on the server – plus, it would get very complicated to emulate the entire [CSSStyleDeclaration](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration) object on the server side.

See `engine-server/src/renderer.ts` for other examples of `unsupportedMethod()`.

**Review by nolanlawson — CHANGES_REQUESTED** [review]
Thank you very much for the PR! This is close, but there is some additional work needed to get this PR over the finish line. 🙂

Our CI is a little fussy about forks, so I would recommend running `yarn test && cd packages/@lwc/integration-karma && yarn test` to verify that your changes are working as expected.

File: packages/@lwc/engine-server/src/__tests__/fixtures/style/expected.html (hunk 4 above)
Comment (le-incroyable1-dev) [reply packages/@lwc/engine-server/src/__tests__/fixtures/style/expected.html:4]: So that would mean that inside the getStyle function in `engine-server/src/renderer.ts `  I need to throw the unsupported method error.  Does this mean that I should just remove `__tests__/fixtures/style/` as we don't need getStyle in the engine-server? If not, what am I supposed to do with it?

File: playground/src/modules/x/counter/counter.html (hunk 3 above)
Comment (le-incroyable1-dev) [reply playground/src/modules/x/counter/counter.html:5]: This was accidental, there is supposed to be no change here

File: packages/@lwc/engine-server/src/__tests__/fixtures/style/expected.html (hunk 4 above)
Comment (nolanlawson) [reply packages/@lwc/engine-server/src/__tests__/fixtures/style/expected.html:4]: @le-incroyable1-dev We already have an existing test for unsupported `engine-server` methods. You can add `this.style` there:

https://github.com/salesforce/lwc/blob/fe90773a72d28038677516bc855c0ea149f56b0d/packages/%40lwc/engine-server/src/__tests__/fixtures/methods-unsupported/modules/x/methods-unsupported/methods-unsupported.js#L3-L12

File: packages/@lwc/integration-karma/test/component/LightningElement.style/index.spec.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+it(
+    'should return the CSSStyleDeclaration and make the appropriate changes', 
+    () => {
+    const elm = createElement('x-test', { is: Test });
+    expect(elm.style).not.toBe(null);
+    // CSSStyleDeclaration must be returned and hence not be null
+    elm.style.color = "red";
+    expect(elm.style.color).toEqual("red");
+     // checks whether the color is changed or not
```
Comment (nolanlawson) [inline packages/@lwc/integration-karma/test/component/LightningElement.style/index.spec.js:12]: Let's also check that the color truly changed.

```suggestion
     await Promise.resolve();
     expect(getComputedStyle(elm).color).toBe('rgb(255, 0, 0)');
```

Note you will need to make the function `async` above.

File: packages/@lwc/integration-karma/test/component/LightningElement.style/x/test/test.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+export default class Test extends LightningElement {
+    static renderMode = 'light';
+    
+    // constructor() {
+    //     super();
+    //     this._style = this.style;
+    // }
+
+    @api
+    get test() {
```
Comment (nolanlawson) [inline packages/@lwc/integration-karma/test/component/LightningElement.style/x/test/test.js:12]:
```suggestion
    get thisDotStyle() {
```

Let's use a more descriptive name than `test`.

Comment (le-incroyable1-dev) [reply packages/@lwc/integration-karma/test/component/LightningElement.style/x/test/test.js:12]: thanks, I'll fix all these issues asap

File: packages/@lwc/integration-karma/test/component/LightningElement.style/index.spec.js (hunk 5 above)
Comment (le-incroyable1-dev) [reply packages/@lwc/integration-karma/test/component/LightningElement.style/index.spec.js:12]: The test for changing color is working correctly now, but 3 other integration-karma tests are failing, how to go about this ???
image
image

Comment (nolanlawson) [reply packages/@lwc/integration-karma/test/component/LightningElement.style/index.spec.js:12]: Interesting, I guess we already had some reflection set up for `style` to reflect to the `style` attribute. I'll have to take a look.

Comment (le-incroyable1-dev) [reply packages/@lwc/integration-karma/test/component/LightningElement.style/index.spec.js]: Yep I thought the same, not sure how to fix this though, and I've pushed whatever changes I've done till this point so you can go through it

Comment (nolanlawson) [reply packages/@lwc/integration-karma/test/component/LightningElement.style/index.spec.js]: Sounds good. I will take a look in a bit. Appreciate all your help on this!

**Comment by nolanlawson** [comment]
I took a look and made some fixes. There were a few things:

First off, the `html-properties` tests were just sanity tests, so we can remove `style` there since it's not really an attribute-property reflection anyway, and it's tested elsewhere. Also your PR does not change how `elm.style` works – i.e. getting `style` from outside an element ([which already supports the CSSStyleDeclaration](https://stackblitz.com/edit/salesforce-lwc-o6y5ra?file=src%2Fmain.js,src%2Fmodules%2Fx%2Fapp%2Fapp.html&title=LWC%20playground)).

Second, it turns out `@api style` is disallowed in the Babel compiler:

https://github.com/salesforce/lwc/blob/ccd0253991b6e5b806369d4e021833ab14c505e5/packages/%40lwc/babel-plugin-component/src/constants.ts#L23-L26

I decided to keep this behavior and add some more tests to assert it. I don't know the reason why we are blocking this, but I don't see a compelling reason to change it.

Finally, I just added some more tests to assert that `this.style` behaves the same as `elm.style`.

**Review by nolanlawson — APPROVED** [review]

**Review by nolanlawson — CHANGES_REQUESTED** [review]
We discussed this internally and realized that this could be a breaking change (/ht @jye-sf). If a component author is doing something like this:

```js
export default class extends LightningElement {
  renderedCallback() {
    this.style = this.style || 'foo'
  }
}
```

... where they expect `this.style` to be undefined at first, then their code will break.

Interestingly we had this same problem with `refs` (which we handled by not setting `this.refs` if there is no `lwc:ref` in the HTML template) and `tagName` (which we overlooked, but didn't seem to break anyone). But `style` is a much more common property name than `tagName` so it's inherently riskier.

I think we may need to do this through API versioning. I will make those changes to this PR, but this also means that this will have to wait until the LWC v7 release.

**Comment by nolanlawson** [comment]
BTW thank you again for your help with this! We will try to get this in by LWC v7 (likely Winter '25).

**Comment by nolanlawson** [comment]
Closing in favor of #4044 since I'm having trouble getting the Sauce Labs runs to work from a forked PR. Thanks again for the contribution!
